### PR TITLE
[threads] switch to GC Safe around pthread calls in mono_thread_internal_set_priority

### PR DESCRIPTION
**Pull request by lambdageek**
Observed pthread_getschedparam taking a lock on OSX

### CODE REVIEW

**Comment by vargaz** [comment]
Is it ok to take a lock if its only taken for a bounded amount of time ?

**Comment by lambdageek** [comment]
@vargaz yes, that's fine as long as the lock is taken consistently either from GC Safe or from GC Unsafe thread state.

The issue is when a thread in GC Safe and another in GC Unsafe thread state both want the same lock. 
 If the GC Safe (preemptively-suspending) thread is preemptively suspended while holding a lock and if the other thread attempts to take the same lock while in GC Unsafe (cooperatively-suspending) thread state, we will get a deadlock between the suspend initiator and the 2 threads contending the lock.

**Comment by lambdageek** [comment]
@monojenkins backport to 2018-08

File: mono/metadata/threads.c [hunk 1]
```diff
@@ -670,12 +670,16 @@ mono_thread_internal_set_priority (MonoInternalThread *internal, MonoThreadPrior
 
 #ifdef HOST_WIN32
 	BOOL res;
+	DWORD last_error;
 
 	g_assert (internal->native_handle);
 
+	MONO_ENTER_GC_SAFE;
 	res = SetThreadPriority (internal->native_handle, priority - 2);
+	last_error = GetLastError ();
```
Comment (jaykrell) [inline mono/metadata/threads.c]: One could avoid the call if res, but it is incredibly cheap, so ok.
SetThreadPriority is a "slow" syscall, GetLastError is not.

We should have ```inline int-or-enum  MonoToWin32ThreadPriority(int-or-enum a) { return a - 2;}```

I added a comment about the `2` somewhere, probably in `cxx`.

**Pull request merged**